File: deploy-client.py
import subprocess, os, sys, json

WSL_GO     = "/usr/local/go/bin/go"
WSL_GOROOT = "/usr/local/go"
WSL_GOPATH = "/home/test/go"
WSL_SRC    = f"{WSL_GOPATH}/src/aly/client/aly-client"

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
SRC_LINUX  = "/mnt/e/Project2026/aly/client/aly-client"

CONFIG_PATH = os.path.join(
    os.environ.get("LOCALAPPDATA", ""),
    "AlyPublish", "config.json"
)

def wsl(cmd):
    r = subprocess.run(["wsl", "bash", "-c", cmd], capture_output=True, text=True, timeout=60)
    if r.returncode != 0 and r.stderr.strip():
        print(f"  [WSL] {r.stderr.strip()}")
    return r

def win_to_wsl(win_path):
    """Convert Windows absolute path (E:\foo\bar) to WSL mount path (/mnt/e/foo/bar)."""
    drive = win_path[0].lower()
    rest = win_path[2:].replace("\\", "/")
    return f"/mnt/{drive}{rest}"

def load_deploy_targets():
    """Read config.json, return deduplicated list of (win_path, wsl_path) deploy targets.
    
    For each project, two targets are produced:
      1) ProjectPath itself
      2) ProjectPath/../UpdateFolder
    """
    if not os.path.exists(CONFIG_PATH):
        print(f"  [WARN] config.json not found: {CONFIG_PATH}")
        return []

    with open(CONFIG_PATH, "r", encoding="utf-8") as f:
        projects = json.load(f)

    targets = []
    seen = set()
    for proj in projects:
        project_path = proj.get("ProjectPath", "")
        if not project_path or not os.path.isabs(project_path):
            continue

        # Target 1: ProjectPath itself
        if project_path not in seen:
            seen.add(project_path)
            targets.append((project_path, win_to_wsl(project_path)))

        # Target 2: ProjectPath/../UpdateFolder
        update_folder = os.path.normpath(os.path.join(project_path, "..", "UpdateFolder"))
        if update_folder not in seen:
            seen.add(update_folder)
            targets.append((update_folder, win_to_wsl(update_folder)))

    return targets

# ── Step 1: Copy to WSL ────────────────────────
print("=" * 50)
print("  [1/4] Copy client to WSL ...")
wsl(f"rm -rf {WSL_SRC}")
wsl(f"mkdir -p {WSL_GOPATH}/src/aly/client")
r = wsl(f"cp -r {SRC_LINUX} {WSL_SRC}")
if r.returncode != 0:
    sys.exit(1)
print("  Done")
print()

# ── Step 2: Build (XP 32-bit) ──────────────────
print("  [2/4] Building aly-client.exe (GOARCH=386) ...")
build = (
    f"cd {WSL_SRC} && "
    f"GOROOT={WSL_GOROOT} GOPATH={WSL_GOPATH} GOOS=windows GOARCH=386 "
    f"{WSL_GO} build -ldflags=\"-s -w\" -o aly-client.exe ."
)
r = wsl(build)
if r.returncode != 0:
    sys.exit(1)
print("  Build OK")
print()

# ── Step 3: Deploy to all targets ──────────────
targets = load_deploy_targets()
if not targets:
    print("  [3/4] No deploy targets found in config.json")
    print()
else:
    print(f"  [3/4] Deploying to {len(targets)} target(s) ...")
    fail = False
    for idx, (win_path, wsl_path) in enumerate(targets, 1):
        print(f"         [{idx}/{len(targets)}] {win_path}")
        r = wsl(f"mkdir -p {wsl_path}")
        if r.returncode != 0:
            print(f"         [FAIL] mkdir failed")
            fail = True
            continue
        r = wsl(f"cp -f {WSL_SRC}/aly-client.exe {wsl_path}/")
        if r.returncode != 0:
            print(f"         [FAIL] copy failed")
            fail = True
            continue
        # show file size
        r = wsl(f"stat -c%s {wsl_path}/aly-client.exe")
        if r.returncode == 0:
            size = int(r.stdout.strip())
            print(f"         OK ({size / 1024 / 1024:.1f} MB)")
        else:
            print(f"         OK")
    if fail:
        sys.exit(1)
    print()

# ── Step 4: Clean ──────────────────────────────
print("  [4/4] Clean WSL ...")
wsl(f"rm -rf {WSL_SRC}")
print("  Done")
print()

print("=" * 50)
if targets:
    for win_path, _ in targets:
        print(f"  Deployed: {win_path}\\aly-client.exe")
else:
    print("  No targets deployed (check config.json)")
print("=" * 50)

File: deploy-client.go
// deploy-client: 通过 WSL 编译 aly-client.exe (32位/XP兼容) 并部署到目标目录
//
// 用法: go run deploy-client.go
// 前置: Windows 已安装 WSL，WSL 内已安装 Go 1.10

package main

import (
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)

const (
	wslGo     = "/usr/local/go/bin/go"
	wslGoroot = "/usr/local/go"
	wslGopath = "/home/test/go"
)

var (
	scriptDir string
	wslSrc    string // WSL 内的源码路径
	linuxSrc  string // 源码在 WSL 中的挂载路径
)

type deployTarget struct {
	win string
	wsl string
}

type projectConfig struct {
	ProjectPath string `json:"ProjectPath"`
}

func mustInit() {
	var err error
	scriptDir, err = os.Getwd()
	if err != nil {
		scriptDir, _ = os.Getwd()
	}
	if scriptDir == "" {
		scriptDir = "."
	}
	scriptDir, _ = filepath.Abs(scriptDir)

	wslSrc = wslGopath + "/src/aly/client/aly-client"
	linuxSrc = "/mnt/e/Project2026/aly/client/aly-client"
}

func wsl(cmd string) error {
	c := exec.Command("wsl", "bash", "-c", cmd)
	c.Stdout = os.Stdout
	c.Stderr = os.Stderr
	return c.Run()
}

func wslOutput(cmd string) (string, error) {
	c := exec.Command("wsl", "bash", "-c", cmd)
	c.Stderr = os.Stderr
	out, err := c.Output()
	return strings.TrimSpace(string(out)), err
}

func winToWsl(winPath string) string {
	if len(winPath) < 2 {
		return winPath
	}
	drive := strings.ToLower(string(winPath[0]))
	rest := strings.ReplaceAll(winPath[2:], "\\", "/")
	return "/mnt/" + drive + rest
}

func configPath() string {
	appdata := os.Getenv("LOCALAPPDATA")
	if appdata == "" {
		home, _ := os.UserHomeDir()
		appdata = filepath.Join(home, "AppData", "Local")
	}
	return filepath.Join(appdata, "AlyPublish", "config.json")
}

func loadDeployTargets() []deployTarget {
	cp := configPath()
	data, err := os.ReadFile(cp)
	if err != nil {
		fmt.Printf("  [WARN] config.json not found: %s\n", cp)
		return nil
	}
	var projects []projectConfig
	if err := json.Unmarshal(data, &projects); err != nil {
		fmt.Printf("  [WARN] config.json parse error: %v\n", err)
		return nil
	}

	targets := []deployTarget{}
	seen := make(map[string]bool)

	for _, proj := range projects {
		pp := proj.ProjectPath
		if pp == "" || !filepath.IsAbs(pp) {
			continue
		}
		pp = filepath.Clean(pp)

		// Target 1: ProjectPath itself
		if !seen[pp] {
			seen[pp] = true
			targets = append(targets, deployTarget{pp, winToWsl(pp)})
		}

		// Target 2: ProjectPath/../UpdateFolder
		updateFolder := filepath.Clean(filepath.Join(pp, "..", "UpdateFolder"))
		if !seen[updateFolder] {
			seen[updateFolder] = true
			targets = append(targets, deployTarget{updateFolder, winToWsl(updateFolder)})
		}
	}
	return targets
}

func main() {
	mustInit()
	fmt.Println(strings.Repeat("=", 50))

	// ── Step 1: Copy to WSL ──
	fmt.Println("  [1/4] Copy client to WSL ...")
	if err := wsl("rm -rf " + wslSrc); err != nil {
		fmt.Printf("  [WARN] rm failed: %v\n", err)
	}
	if err := wsl("mkdir -p " + wslGopath + "/src/aly/client"); err != nil {
		fmt.Printf("  [ERROR] mkdir failed: %v\n", err)
		os.Exit(1)
	}
	if err := wsl("cp -r " + linuxSrc + " " + wslSrc); err != nil {
		fmt.Printf("  [ERROR] copy to WSL failed: %v\n", err)
		os.Exit(1)
	}
	fmt.Println("  Done\n")

	// ── Step 2: Build (XP 32-bit) ──
	fmt.Println("  [2/4] Building aly-client.exe (GOARCH=386) ...")
	buildCmd := fmt.Sprintf(
		"cd %s && GOROOT=%s GOPATH=%s GOOS=windows GOARCH=386 %s build -ldflags=\"-s -w\" -o aly-client.exe .",
		wslSrc, wslGoroot, wslGopath, wslGo,
	)
	if err := wsl(buildCmd); err != nil {
		fmt.Printf("  [ERROR] Build failed: %v\n", err)
		os.Exit(1)
	}
	fmt.Println("  Build OK\n")

	// ── Step 3: Deploy to all targets ──
	targets := loadDeployTargets()
	if len(targets) == 0 {
		fmt.Println("  [3/4] No deploy targets found in config.json\n")
	} else {
		fmt.Printf("  [3/4] Deploying to %d target(s) ...\n", len(targets))
		fail := false
		for i, t := range targets {
			fmt.Printf("         [%d/%d] %s\n", i+1, len(targets), t.win)
			if err := wsl("mkdir -p " + t.wsl); err != nil {
				fmt.Printf("         [FAIL] mkdir failed: %v\n", err)
				fail = true
				continue
			}
			if err := wsl("cp -f " + wslSrc + "/aly-client.exe " + t.wsl + "/"); err != nil {
				fmt.Printf("         [FAIL] copy failed: %v\n", err)
				fail = true
				continue
			}
			out, err := wslOutput("stat -c%s " + t.wsl + "/aly-client.exe")
			if err == nil {
				var size int64
				fmt.Sscanf(out, "%d", &size)
				fmt.Printf("         OK (%.1f MB)\n", float64(size)/1024/1024)
			} else {
				fmt.Println("         OK")
			}
		}
		if fail {
			os.Exit(1)
		}
		fmt.Println()
	}

	// ── Step 4: Clean ──
	fmt.Println("  [4/4] Clean WSL ...")
	if err := wsl("rm -rf " + wslSrc); err != nil {
		fmt.Printf("  [WARN] clean failed: %v\n", err)
	}
	fmt.Println("  Done\n")

	fmt.Println(strings.Repeat("=", 50))
	if len(targets) > 0 {
		for _, t := range targets {
			fmt.Printf("  Deployed: %s\\aly-client.exe\n", t.win)
		}
	} else {
		fmt.Println("  No targets deployed (check config.json)")
	}
	fmt.Println(strings.Repeat("=", 50))
}
